content/updates/understanding-the-terraform-check-block-feature.md

```diff
@@ -0,0 +1,207 @@
+
+Also, this allows to utilize [`external`](https://registry.terraform.io/providers/hashicorp/external/latest/docs/data-sources/external) data source provided by Hashicorp. It green-lights integration of arbitrary external scripts into your Terraform configuration. This can be a Python script, a shell script, or any other program that can read JSON from standard input and write JSON to standard output. It's worth noting that the use of the external data source comes with a caveat: it can potentially make your configuration depend on the specific environment where it runs, as it might rely on specific external scripts and language runtimes being available.
+
+Let's consider an example from Masterpoint's module [masterpointio/terraform-aws-tailscale](https://github.com/masterpointio/terraform-aws-tailscale/tree/feature/check_auth_key) where we validate the desired object state using a data source. The module provisions an AWS EC2 instance that serves as a [Tailscale subnet router](https://tailscale.com/kb/1019/subnets/) device. The subnet router setup is deployed via `userdata`, and upon authentication with a [Tailnet key](https://registry.terraform.io/providers/tailscale/tailscale/latest/docs/resources/tailnet_key) — managed by Terraform — the device is automatically tagged. Correct tagging of the Tailscale subnet router is critical for its proper functioning. We can introduce the checks in [checks.tf](https://github.com/masterpointio/terraform-aws-tailscale/blob/feature/check_auth_key/checks.tf) to validate if the tags were successfully applied.
+
+The first assertions verifies that the device has been tagged; the second one confirms that the tags applied are as expected:
+
+```hcl
+variable "tailnet_name" {
+  type        = string
+  description = <<EOF
+  This unique name of the Tailnet that is used when registering DNS entries, e.g. 'cat-crocodile.ts.net'.
+  See https://tailscale.com/kb/1217/tailnet-name/ for more information.
+  EOF
+}
+
+check "device" {
+  data "tailscale_device" "default" {
+    name     = format("%s.%s", module.this.id, var.tailnet_name)
+    wait_for = "30s"
+  }
+
+  assert {
+    condition     = length(data.tailscale_device.default.tags) > 0
+    error_message = "Device ${data.tailscale_device.default.name} is not tagged."
+  }
+
+  assert {
+    condition     = sort(data.tailscale_device.default.tags) == sort(tolist(local.tailscale_tags))
+    error_message = <<EOF
+    Device ${data.tailscale_device.default.name} is not tagged with the correct tags.
+    The list of expected tags is: [${join(",", local.tailscale_tags)}]
+    EOF
+  }
+}
+```
+
+Example of the failed assertion:
+```sh
+Warning: Check block assertion failed
+  on ../../../terraform-aws-tailscale/checks.tf line 19, in check "device":
+  19:     condition     =  sort(data.tailscale_device.default.tags) == sort(tolist(local.tailscale_tags))
+    ├────────────────
+    │ data.tailscale_device.default.tags is set of string with 1 element
```

Comment:
This example failure is frustrating cause it's just the type and not the elements that are not matching up. Anything we can do about that?

Reply:
TBH, I don't know why TF prints this info about the types. Yeah, initially, these are different types, but I use `tolist` and `sort` functions here to compare the tags correctly.
In this test, I just add some extra symbols to local tags, and that causes failure, which is expected. We can add a message like this:
```sh
│ Device mp-automation-tailscale-subnet-router.cat-crocodile.ts.net is not tagged with the correct tags.
│     The list of expected tags is: ["tag:mp-automation-tailscale-subnet-router-test"]
```
Does that look better? Or should I look for another less confusing test?

Reply:
That's great -- Thanks! 👍